src/15_hello_triangle.rs

```diff
@@ -0,0 +1,776 @@
+extern crate env_logger;
+#[cfg(feature = "dx12")]
+extern crate gfx_backend_dx12 as back;
+#[cfg(feature = "metal")]
+extern crate gfx_backend_metal as back;
+#[cfg(feature = "vulkan")]
+extern crate gfx_backend_vulkan as back;
+extern crate gfx_hal as hal;
+extern crate glsl_to_spirv;
+extern crate winit;
+
+use hal::{
+    Capability, Device, Instance, PhysicalDevice, QueueFamily, Surface, Swapchain, SwapchainConfig,
+};
+use std::io::Read;
+
+static WINDOW_NAME: &str = "15_command_buffers";
+
+const MAX_FRAMES_IN_FLIGHT: usize = 2;
+
+fn main() {
+    env_logger::init();
+    let (window, events_loop) = init_window();
+
+    let (
+        _instance,
+        device,
+        mut command_queues,
+        mut swapchain,
+        _surface,
+        frame_images,
+        render_pass,
+        descriptor_set_layouts,
+        pipeline_layout,
+        gfx_pipeline,
+        swapchain_framebuffers,
+        command_pool,
+        submission_command_buffers,
+        image_available_semaphores,
+        render_finished_semaphores,
+        in_flight_fences,
+    ) = init_hal(&window);
+
+    main_loop(
+        &device,
+        events_loop,
+        &mut command_queues,
+        &mut swapchain,
+        &submission_command_buffers,
+        &image_available_semaphores,
+        &render_finished_semaphores,
+        &in_flight_fences,
+    );
+
+    clean_up(
+        device,
+        frame_images,
+        swapchain,
+        render_pass,
+        descriptor_set_layouts,
+        pipeline_layout,
+        gfx_pipeline,
+        swapchain_framebuffers,
+        command_pool,
+        vec![image_available_semaphores, render_finished_semaphores],
+        in_flight_fences,
+    );
+}
+
+fn draw_frame(
+    device: &<back::Backend as hal::Backend>::Device,
+    command_queues: &mut Vec<hal::queue::CommandQueue<back::Backend, hal::Graphics>>,
+    swapchain: &mut <back::Backend as hal::Backend>::Swapchain,
+    submission_command_buffers: &Vec<
+        hal::command::Submit<
+            back::Backend,
+            hal::Graphics,
+            hal::command::MultiShot,
+            hal::command::Primary,
+        >,
+    >,
+    image_available_semaphore: &<back::Backend as hal::Backend>::Semaphore,
+    render_finished_semaphore: &<back::Backend as hal::Backend>::Semaphore,
+    in_flight_fence: &<back::Backend as hal::Backend>::Fence,
+) {
+    device.wait_for_fence(in_flight_fence, std::u64::MAX);
+    device.reset_fence(in_flight_fence);
+
+    let image_index = swapchain
+        .acquire_image(
+            std::u64::MAX,
+            hal::window::FrameSync::Semaphore(image_available_semaphore),
+        ).expect("could not acquire image!");
+
+    // Submission::new creates only a transfer capable submission --- how do things end up working out so that we can submit graphics commands?
```

Comment:
Which part are you wondering about? Maybe I can clarify

Reply:
`Submission::new` creates (per the docs) "Create a new empty (transfer) submission." However, we are submitting graphics submits (full type: `hal::command::Submit<back::Backend, hal::Graphics, hal::command::MultiShot, hal::command::Primary,>`). How does the capability change of Submission (from transfer to graphics) happen when we provide a graphics submit?

Reply:
See this comment on `submit`:
> Submission will be automatically promoted to to the minimum required capability to hold all passed submits.